rectools/models/nn/transformer_data_preparator.py

```diff
@@ -0,0 +1,372 @@
+from collections.abc import Hashable
+
+import numpy as np
+import pandas as pd
+import torch
+from scipy import sparse
+from torch.utils.data import DataLoader
+from torch.utils.data import Dataset as TorchDataset
+
+from rectools import Columns, ExternalIds
+from rectools.dataset import Dataset, Interactions
+from rectools.dataset.features import SparseFeatures
+from rectools.dataset.identifiers import IdMap
+
+from .constants import PADDING_VALUE
+
+
+class SequenceDataset(TorchDataset):
+    """
+    Dataset for sequential data.
+
+    Parameters
+    ----------
+    sessions : List[List[int]]
+        User sessions in the form of sequences of items ids.
+    weights : List[List[float]]
+        Weight of each interaction from the session.
+    """
+
+    def __init__(self, sessions: tp.List[tp.List[int]], weights: tp.List[tp.List[float]]):
+        self.sessions = sessions
+        self.weights = weights
+
+    def __len__(self) -> int:
+        return len(self.sessions)
+
+    def __getitem__(self, index: int) -> tp.Tuple[tp.List[int], tp.List[float]]:
+        session = self.sessions[index]  # [session_len]
+        weights = self.weights[index]  # [session_len]
+        return session, weights
+
+    @classmethod
+    def from_interactions(
+        cls,
+        interactions: pd.DataFrame,
+        sort_users: bool = False,
+    ) -> "SequenceDataset":
+        """
+        Group interactions by user.
+        Construct SequenceDataset from grouped interactions.
+
+        Parameters
+        ----------
+        interactions : pd.DataFrame
+            User-item interactions.
+        """
+        sessions = (
+            interactions.sort_values(Columns.Datetime, kind="stable")
+            .groupby(Columns.User, sort=sort_users)[[Columns.Item, Columns.Weight]]
+            .agg(list)
+        )
+        sessions, weights = (
+            sessions[Columns.Item].to_list(),
+            sessions[Columns.Weight].to_list(),
+        )
+
+        return cls(sessions=sessions, weights=weights)
+
+
+class TransformerDataPreparatorBase:
+    """
+    Base class for data preparator. To change train/recommend dataset processing, train/recommend dataloaders inherit
+    from this class and pass your custom data preparator to your model parameters.
+
+    Parameters
+    ----------
+    session_max_len : int
+        Maximum length of user sequence.
+    batch_size : int
+        How many samples per batch to load.
+    dataloader_num_workers : int
+        Number of loader worker processes.
+    item_extra_tokens : Sequence(Hashable)
+        Which element to use for sequence padding.
+    shuffle_train : bool, default True
+        If ``True``, reshuffles data at each epoch.
+    train_min_user_interactions : int, default 2
+        Minimum length of user sequence. Cannot be less than 2.
+    get_val_mask_func : Callable, default None
+        Function to get validation mask.
+    """
+
+    train_session_max_len_addition: int = 0
```

Comment:
We sometimes need data preparators to add +1 to actual session_max_len. SASRec needs it for shifted sequence. That's why it's here.

Reply:
Let's put this comment there

Reply:
done